src/sbvr-api/translations.ts

```diff
@@ -0,0 +1,215 @@
+		if (mapping != null && mapping.length === 2) {
+			if (!key.includes('$')) {
+				mapping = _.cloneDeep(mapping);
+				mapping[1]![0] = `${mapping[1]![0]}$${alias}`;
+				(relationships as RelationshipInternalNode)[`${key}$${alias}`] = {
+					$: mapping,
+				};
+				delete (relationships as RelationshipInternalNode)[key];
+			}
+		}
+		namespaceRelationships(relationship, alias);
+	});
+};
+
+export const translateAbstractSqlModel = (
+	fromAbstractSqlModel: AbstractSqlModel,
+	toAbstractSqlModel: AbstractSqlModel,
+	fromVersion: string,
+	toVersion: string,
+	definitions: _.Dictionary<
+		| (Definition & { $toResource?: string })
+		| _.Dictionary<string | ReferencedFieldNode>
+	> = {},
+): _.Dictionary<string> => {
+	const resourceRenames: _.Dictionary<string> = {};
+
+	// TODO: why?
+	fromAbstractSqlModel.rules = toAbstractSqlModel.rules;
+
+	const fromKeys = Object.keys(fromAbstractSqlModel.tables);
+	const nonexistentTables = _.difference(Object.keys(definitions), fromKeys);
+	if (nonexistentTables.length > 0) {
+		throw new Error(
+			`Tried to define non-existent resources: '${nonexistentTables.join(
+				', ',
+			)}'`,
+		);
+	}
+	_.forEach(toAbstractSqlModel.synonyms, (canonicalForm, synonym) => {
+		// Don't double alias
+		if (synonym.includes('$')) {
+			fromAbstractSqlModel.synonyms[synonym] = canonicalForm;
+		} else {
+			fromAbstractSqlModel.synonyms[
+				`${synonym}$${toVersion}`
+			] = `${canonicalForm}$${toVersion}`;
+		}
+	});
+	const relationships = _.cloneDeep(toAbstractSqlModel.relationships);
+	namespaceRelationships(relationships, toVersion);
+	_.forEach(relationships, (relationship, key) => {
+		// Don't double alias
+		if (!key.includes('$')) {
+			key = `${key}$${toVersion}`;
+		}
+		fromAbstractSqlModel.relationships[key] = relationship;
+	});
+
+	// TODO: We also need to keep the original relationship refs to non $version resources
+	// Also alias for ourselves to allow explicit referencing
+	const aliasedFromRelationships = _.cloneDeep(
+		fromAbstractSqlModel.relationships,
+	);
+	namespaceRelationships(aliasedFromRelationships as Relationship, fromVersion);
+	_.forEach(aliasedFromRelationships, (relationship, key) => {
+		// Don't double alias
+		if (!key.includes('$')) {
+			key = `${key}$${fromVersion}`;
+			fromAbstractSqlModel.relationships[key] = relationship;
+		}
+	});
+
+	_.forEach(toAbstractSqlModel.tables, (table, key) => {
+		// Don't double alias
+		if (!key.includes('$')) {
+			key = `${key}$${toVersion}`;
+		}
+		fromAbstractSqlModel.tables[key] = _.cloneDeep(table);
+	});
+
+	fromKeys.forEach((key) => {
+		const definition = definitions[key];
+		const table = fromAbstractSqlModel.tables[key];
+		if (definition) {
+			const hasToResource = typeof definition.$toResource === 'string';
+			if (hasToResource) {
+				resourceRenames[key] = `${definition.$toResource}`;
+			}
+			const toResource = hasToResource
+				? (definition.$toResource as string)
+				: `${key}$${toVersion}`;
```

Comment:
NIT:
```suggestion
			const { $toResource } = definition;
			const hasToResource = typeof $toResource === 'string';
			if (hasToResource) {
				resourceRenames[key] = $toResource;
			}
			const toResource = hasToResource
				? $toResource
				: `${key}$${toVersion}`;
```
It seems like a TS limitation for the aliased condition when it comes from an object property, given that if we did `if (typeof definition.$toResource === 'string')` it would work as expected.
See: https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#control-flow-analysis-for-bracketed-element-access